### DISCUSSION TITLE
Controlling two Yaskawa Robots via Moveit

### DISCUSSION THREAD

**Discussion by TE-2** [Yaskawa-Global/motoros2#225] [Q&A] [answered]
Hello everyone,
I have two Yaskawa robots, MotoMini and GP4, in a robot module and would like to control them simultaneously via ROS2. The control of a single robot via Moveit works without problems, I used the [ros2-starter-for-yaskawa-robots repository](https://github.com/YaskawaEurope/ros2-starter-for-yaskawa-robots/tree/main) and adapted the ros2 package [motoXY ](https://github.com/YaskawaEurope/ros2-starter-for-yaskawa-robots/tree/main/src/motoXY)to my two robots.

For the control of both robots I want a move_group node for each robot in a respective namespace e.g. (`moto_mini_moveit `& `moto_gp4_moveit`), because I read in a [forum post](https://answers.ros.org/question/362796/how-to-simulate-and-control-2-industrial-robots-with-moveit/) that the move group node is not able to execute two trajectories at the same time.

For this I have modified [xy_start.launch.py](https://github.com/YaskawaEurope/ros2-starter-for-yaskawa-robots/blob/main/src/motoXY/moto_moveit_config/launch/xy_start.launch.py) by passing `run_move_group_nodeX` the new name_space as an argument, as well as `joint_state_publisherX` and `robot_state_publisherX` and would like to start only these three nodes for the time being.
With the help of rqt_graph I was able to identify the incorrectly assigned topics and then wrote the suitable remappings. The direct mapping of actions is unfortunately not yet possible according to [this article](https://github.com/ros2/ros2/issues/1312#issuecomment-1705521109), so I used the workaround and remapped the corresponding topics.

The node `/moveit_simple_controller_manager`, for example, even had `/moto_mini_moveit/moto_mini/follow_joint_trajectory/_action/status` after the action topic. 

Customised `xy_start.launch.py`:

<details><summary>Click to expand</summary>

```python
import os
from launch import LaunchDescription
from launch_ros.actions import Node
from ament_index_python.packages import get_package_share_directory
from moveit_configs_utils import MoveItConfigsBuilder

import xacro

def generate_launch_description():
# =======================================================================
# NameSpace configuration
    
    robot_name_space        = "/moto_mini"
    name_space_moveit       = "/moto_mini_moveit"

# =======================================================================
    motoman_directory = "moveit_resources_moto_motomini_moveit_config"  #not Directory -> ROS Package
# URDF
    motoman_robot_description_config = xacro.process_file(
        os.path.join(
            get_package_share_directory(motoman_directory),
            "config",
            "motoman_motomini.urdf.xacro",
        ),
    )
    motoman_robot_description = {
        "robot_description": motoman_robot_description_config.toxml()
    }
# =======================================================================
# MoveGroupNode
    moveit_config = (
        MoveItConfigsBuilder("moveit_resources_moto_motomini")
        .robot_description(file_path="config/motoman_motomini.urdf.xacro")
        .robot_description_semantic(file_path="config/motoman_motomini.srdf")
        .trajectory_execution(file_path="config/moveit_controllers.yaml")
        .to_moveit_configs()
    )
    planning_scene_monitor_parameters = {
        "publish_planning_scene": True,
        "publish_geometry_updates": True,
        "publish_state_updates": True,
        "publish_transforms_updates": True,
        "publish_robot_description": True,
        "publish_robot_description_semantic": True,
    }
    # Start the actual move_group node/action server
    # robot
    run_move_group_nodeX = Node(
        package="moveit_ros_move_group",
        executable="move_group",
        namespace= name_space_moveit,
        output="screen",
        parameters=[
            moveit_config.to_dict(),
            planning_scene_monitor_parameters,
        ],
        remappings= [
            ('joint_states',                                                                      f'{robot_name_space}/joint_states'),               
            # Action -> follow_joint_trajectory action
            (f'{name_space_moveit}{robot_name_space}/follow_joint_trajectory/_action/status',     f'{robot_name_space}/follow_joint_trajectory/_action/status'),
            (f'{name_space_moveit}{robot_name_space}/follow_joint_trajectory/_action/feedback',   f'{robot_name_space}/follow_joint_trajectory/_action/feedback'),

            (f'{name_space_moveit}/monitored_planning_scene',                                      '/monitored_planning_scene'),
            (f'{name_space_moveit}/display_planned_path',                                          '/display_planned_path'),
        ]
    )
# =======================================================================
    joint_state_publisherX = Node(
        package='joint_state_publisher',
        executable='joint_state_publisher',
        name='joint_state_publisher',
        namespace= name_space_moveit,
        parameters=[
            {
                "robot_description": motoman_robot_description,
                "rate": 43,
                "source_list": [f'{robot_name_space}/joint_states'],
            }
        ],
    )
# =======================================================================
    robot_state_publisherX = Node(
        package="robot_state_publisher",
        executable="robot_state_publisher",
        name="robot_state_publisher",
        namespace= name_space_moveit,
        parameters=[
            motoman_robot_description,
            {
                "robot_description": motoman_robot_description,
                "publish_frequency": 43.0,
                "frame_prefix": "",
            },
        ]
    )
# =======================================================================
# Launch
    return LaunchDescription([
            run_move_group_nodeX,
            joint_state_publisherX,
            robot_state_publisherX,
            ]
    )
```

</details>

After this adjustment my ros graph looked correct again but with the new namespace:

<details><summary>Click to expand</summary>

![rosgraph_namespace](https://github.com/Yaskawa-Global/motoros2/assets/84192658/3b7f5487-bd0a-4071-912b-3914bec7ac10)

</details>

Now I wanted to test it using a script of the [hello_moveit package](https://github.com/YaskawaEurope/ros2-starter-for-yaskawa-robots/blob/main/src/hello_moveit/src/hello_moveit.cpp).
For this it was important to add options to the constructor of the MoveGroupInterface in which the move_group_namespace was defined.

```C++
std::string group_name = "manipulator";
std::string move_group_namespace = "/moto_mini_moveit";
moveit::planning_interface::MoveGroupInterface::Options options(group_name, moveit::planning_interface::MoveGroupInterface::ROBOT_DESCRIPTION, move_group_namespace);
using moveit::planning_interface::MoveGroupInterface;
auto move_group_interface = MoveGroupInterface(node, options);
```

In addition, I had to start the node via the following launch file in which I again specified the new namespace and the mapping had to be adapted. Without the launch file and running the script via ros2 run I could not even load the robot model even if I gave the argument namespace "moto_mini_moveit" to the constructor of the "hello_motomini" node.

<details><summary>Click to expand</summary>

```python
from launch import LaunchDescription
from launch_ros.actions import Node
def generate_launch_description():
    name_space_moveit   = "/moto_mini_moveit"

    return LaunchDescription([
        Node(
            package='hello_moveit',
            executable='hello_motomini',
            name='hello_motomini',
            namespace= name_space_moveit,
            output="screen",
            remappings= [
                # Action -> move_action
                ('/move_action/_action/feedback',           f'{name_space_moveit}/move_action/_action/feedback'),
                ('/move_action/_action/status',             f'{name_space_moveit}/move_action/_action/status'),
                # Action -> execute_trajectory ction
                ('/execute_trajectory/_action/feedback',    f'{name_space_moveit}/execute_trajectory/_action/feedback'),
                ('/execute_trajectory/_action/status',      f'{name_space_moveit}/execute_trajectory/_action/status'),

                ('/trajectory_execution_event',             f'{name_space_moveit}/trajectory_execution_event'),
                ('/attached_collision_object',              f'{name_space_moveit}/attached_collision_object'),
            ]
        )
    ])
```

</details>

With the help of the launch file, the node started correctly and the planning was also executed. Only the execution of the trajectory fails and I don't understand why.

In the Terminal of` xy_start.launch` I get the following logs:

<details><summary>Click to expand</summary>

```
[move_group-1] You can start planning now!
[move_group-1] 
[joint_state_publisher-2] [INFO] [1710347636.450236976] [moto_mini_moveit.joint_state_publisher]: Waiting for robot_description to be published on the robot_description topic...
[move_group-1] [INFO] [1710349284.843051471] [moveit_move_group_default_capabilities.move_action_capability]: Received request
[move_group-1] [INFO] [1710349284.843127327] [moveit_move_group_default_capabilities.move_action_capability]: executing..
[move_group-1] [INFO] [1710349284.869346679] [moveit_move_group_default_capabilities.move_action_capability]: Planning request received for MoveGroup action. Forwarding to planning pipeline.
[move_group-1] [INFO] [1710349284.870056364] [moveit.ompl_planning.model_based_planning_context]: Planner configuration 'manipulator' will use planner 'geometric::RRTConnect'. Additional configuration parameters will be set when the planner is constructed.
[move_group-1] [WARN] [1710349284.900956562] [moveit_trajectory_processing.time_optimal_trajectory_generation]: Joint acceleration limits are not defined. Using the default 1 rad/s^2. You can define acceleration limits in the URDF or joint_limits.yaml.
[move_group-1] [INFO] [1710349284.906044218] [moveit_move_group_default_capabilities.move_action_capability]: Motion plan was computed successfully.
[move_group-1] [INFO] [1710349288.907850779] [moveit_move_group_default_capabilities.execute_trajectory_action_capability]: Received goal request
[move_group-1] [INFO] [1710349288.907909092] [moveit_move_group_default_capabilities.execute_trajectory_action_capability]: Execution request received
[move_group-1] [INFO] [1710349288.907929054] [moveit.plugins.moveit_simple_controller_manager]: Returned 1 controllers in list
[move_group-1] [INFO] [1710349288.907937800] [moveit.plugins.moveit_simple_controller_manager]: Returned 1 controllers in list
[move_group-1] [INFO] [1710349288.907985322] [moveit_ros.trajectory_execution_manager]: Validating trajectory with allowed_start_tolerance 0.01
[move_group-1] [INFO] [1710349288.920418642] [moveit_ros.trajectory_execution_manager]: Starting trajectory execution ...
[move_group-1] [INFO] [1710349288.920457918] [moveit.plugins.moveit_simple_controller_manager]: Returned 1 controllers in list
[move_group-1] [INFO] [1710349288.920464835] [moveit.plugins.moveit_simple_controller_manager]: Returned 1 controllers in list
[move_group-1] [ERROR] [1710349288.920511311] [moveit.simple_controller_manager.follow_joint_trajectory_controller_handle]: Action client not connected to action server: moto_mini/follow_joint_trajectory
[move_group-1] [ERROR] [1710349288.920513743] [moveit_ros.trajectory_execution_manager]: Failed to send trajectory part 1 of 1 to controller moto_mini
[move_group-1] [INFO] [1710349288.920516026] [moveit_ros.trajectory_execution_manager]: Completed trajectory execution with status ABORTED ...
[move_group-1] [INFO] [1710349288.920660570] [moveit_move_group_default_capabilities.execute_trajectory_action_capability]: Execution completed: ABORTED
```

</details>

It seems to fail because of the following error message:

```
[moveit.simple_controller_manager.follow_joint_trajectory_controller_handle]: Action client not connected to action server: moto_mini/follow_joint_trajectory
```

But I don't understand this, because according to rqt_graph and "ros2 node info ..." the `moveit_simple_controller_manager` is a client of `/moto_mini/follow_joint_trajectory/_action` ??? (see ros graph picture above)

I would appreciate any helpful input.
Thanks

**Comment by gavanderhoorn** [comment]
> I have two Yaskawa robots, MotoMini and GP4, in a robot module and would like to control them simultaneously via ROS2.

could you please clarify what you mean by this?

Do you have both a MotoMini and a GP4 controlled by a single YRC1000? Or do you have two controllers, one for each robot?

Could you please also provide more information on which version of MotoROS2 you are using? And which version of ROS 2?

**Reply by TE-2** [reply]
Yes, of course, the two robots each have their own YRC1000micro controller and work in a production module. The robots should not work directly together, but their workspaces overlap. They should also be able to execute trajectories simultaneously so that one robot does not have to wait for the other.

We use version 0.1.2 of motoros2 and the ROS2 distribution Humble.

**Reply by gavanderhoorn** [reply]
Technically, this is most likely a MoveIt 2 problem, and I would recommend you open an issue on their issue tracker: [ros-planning/moveit2](https://github.com/ros-planning/moveit2/issues).

Having written that: I would perhaps suggest to avoid remaps as much as possible, and see if launching the two MoveIt instances you have each in their own namespace works well enough. I don't believe any remaps would be needed any more, as you should be able to configure the MoveIt Simple Controller to use an action server on any topic, regardless of in which namespace it is running itself.

> the two robots each have their own YRC1000micro controller [..]

Could you please attach the contents of the configuration `.yaml` files for each of the controllers? Please remove all the comments, as we're only interested in the key-value pairs.

> The robots should not work directly together, but their workspaces overlap

with two MoveIt instances, you'd probably have to make sure each instance 'knows' about the other robot. Providing both instances with the same planning scene (ie: URDF), but just controlling their own robot should be sufficient, provided the `JointState`s of the 'uncontrolled' robot still reach MoveIt (so it can update its planning scene with the 'current state' of the other robot).

**Reply by gavanderhoorn** [reply]
@TE-2: I just saw your ros-planning/moveit2#2744.

To exclude a few things on the MotoROS2 side, could I ask you to still post your `motoros2_config.yaml`s for your two controllers?

If I understand your setup/description correctly, I would expect the following to be sufficient:

1. start each MotoROS2 node in its own namespace (by [configuring appropriate values for `node_namespace`](https://github.com/Yaskawa-Global/motoros2/blob/5f8c80011c64dc9152a4c72b2eb138760a9d649e/config/motoros2_config.yaml#L36-L40))
2. start each MoveIt instance in the same (ie: its own) namespace (ie: the complete launch file)

I'm most likely not considering everything (especially MoveIt's configuration can be complex), but if the `action_ns` and MoveIt controller names are correctly configured (and use relative names), the two MoveIt instances might even find 'their own' `FollowJointTrajectory` action servers in their own namespaces *automatically* as relative lookups should start in the namespace the `move_group` node is started in (alternatively you could use absolute names of course).

You might not even need any remaps in this case.

But to check the MotoROS2 side, we'd need to see your `motoros2_config.yaml` files.

**Reply by TE-2** [reply]
> Technically, this is most likely a MoveIt 2 problem

Yes, that makes sense. Here is the [link to the new issue](https://github.com/ros-planning/moveit2/issues/2744#issue-2190725481) in the moveit repository (thanks for improving the markdown). 
I just thought maybe someone here in the yaskawa community already had a similar setup or I did something wrong on the yaskawa side.

> Could you please attach the contents of the configuration .yaml files for each of the controllers?

Sure, here are my configurations:
GP4 configuration:
<details><summary>Click to expand</summary>

```yaml
agent_ip_address: xxx.xxx.xxx.xxx
agent_port_number: 8888
ros_domain_id: 0
node_name: "yrc_1000_micro"
node_namespace: "moto_gp4"
sync_timeclock_with_agent: true
stop_motion_on_disconnect: true
publish_tf: false
namespace_tf: true
joint_names:
- [joint_1, joint_2, joint_3, joint_4, joint_5, joint_6]
update_periods:
  executor_sleep_period: 10
  feedback_publisher_period: 20
  controller_io_status_monitor_period: 10
publisher_qos:
  robot_status: sensor_data
  joint_states: default
  tf: default
```
</details>

In the Motomini configuration only the node_namespace was changed to "moto_mini".

> Providing both instances with the same planning scene (ie: URDF)

Yes that was what we had in mind. As a short-term workaround, we set both robots to their own `ROS Domain ID` in order to be able to continue working. Of course this works, but now we cannot share the planning scenes.

**Reply by TE-2** [reply]
yes, I was probably a bit too slow ^^, I wanted to link the issue here and I posted it there first

**Reply by TE-2** [reply]
okay so if I understand correctly the Moveit Launch file should be started in the same namespace as the Yaskawa node so the Move Group Node would be e.g. `/moto_mini/move_group` ?

**Reply by gavanderhoorn** [reply]
> okay so if I understand correctly the Moveit Launch file should be started in the same namespace as the Yaskawa node so the Move Group Node would be e.g. `/moto_mini/move_group` ?

yes, that would be what I would first try.

Remaps would be my last resort.

**Reply by TE-2** [reply]
All right, thanks for the hint, I'll try it out in the next few days.

**Reply by TE-2** [reply]
Hi,
I just remembered why I decided to create a separate namespace for the move group node. Both the Yaskawa Node and Moveit create a topic with the name `/<namespace>/joint_states` (see ROS Graph). When starting moveit in the namespace of the yaskawa node, this leads to a problem because the topics are duplicated.
Is it possible to rename this topic on the yaskawa or moveit site without remapping?

**Reply by gavanderhoorn** [reply]
> Both the Yaskawa Node and Moveit create a topic with the name `/<namespace>/joint_spaces` (see ROS Graph). 

I believe you mean `joint_states`, correct?

If so: MoveIt is a subscriber to that topic, while MotoROS2 publishes to it.

> When starting moveit in the namespace of the yaskawa node, this leads to a problem because the topics are duplicated.

If you are using a MoveIt configuration that publishes to `joint_states`, something isn't right. With a real robot, MoveIt should only consume `JointState` messages, not produce them.

**Reply by gavanderhoorn** [reply]
> Both the Yaskawa Node and Moveit create a topic with the name `/<namespace>/joint_spaces` (see ROS Graph)

The `rqt_graph` screenshot you included shows you have a `joint_state_publisher` running.

Did you include that on purpose (you could fi use it to merge different `joint_state` topics into a single one)? If not, that would be a problem.

And btw:

> this leads to a problem because the topics are duplicated.

it's perfectly OK to have multiple nodes publish to the same topic. That's expressly supported by ROS. Whether consumers know how to process the resulting message stream correctly is something else, but there's nothing inherently problematic about multiple publishers on a single topic.

**Reply by TE-2** [reply]
> I believe you mean joint_states, correct?

yes exactly, I have corrected it

>  The rqt_graph screenshot you included shows you have a joint_state_publisher running.

yes that's right, I have commented out all nodes except `run_move_group_nodeX`, `joint_state_publisherX` and `robot_state_publisherX` in the [xy_start.launch.py](https://github.com/YaskawaEurope/ros2-starter-for-yaskawa-robots/blob/b5a1e15a3a068f46e68d6b5ee90752c778c411b8/src/motoXY/moto_moveit_config/launch/xy_start.launch.py#L217-L219) file. I assumed that these had to be used for the Moveit.

> it's perfectly OK to have multiple nodes publish to the same topic.

good to know!

**Reply by gavanderhoorn** [reply]
I have to admit I haven't used the package you link to.

I might perhaps suggest to start from a more simple setup.

From a quick look it appears that launch file uses a JSP to give MoveIt (which is not running in a namespace) access to the `/yaskawa/joint_states` topic (by using the `source_list` parameter). That basically makes that JSP instance *republish* whatever is published on `/yaskawa/joint_states` to `joint_states` (ie: without the namespace).

But if you run MoveIt itself in the same namespace as you have MotoROS2, I would not expect that to be necessary. Resolution of relative topic names should start in the 'local' namespace.

**Comment by gavanderhoorn** [comment]
@TE-2: have you been able to get things to work?

**Reply by TE-2** [reply]
Hi,
no unfortunately not, I currently have an important deadline to fulfil. I would report back as soon as I was able to try it out.

**Reply by TE-2** [reply]
small update, our robot module will be dispatched in a few days to exhibit it at a fair. We have currently developed it further with the workaround solution of two ROS Domain Ids. After the fair (in 4 Weeks) I will look again at running both Ros Graphs in the same namespace.

**Reply by gavanderhoorn** [reply]
Edit: the comment before this comment was 'promoted' to an answer. See https://github.com/Yaskawa-Global/motoros2/discussions/225#discussioncomment-9859474 below.

---

Thanks for the update and the launch file.

If you'd post this as a new *answer* (below), @TE-2 could select it as *the* answer.

re: `ros2_control` issue: is that purely a problem with loading parameters? I'm not sure what MoveIt would need to interact with the `controller_manager` for (I'd expect it to need to communicate with the controllers/broadcasters, but not the `controller_manager` itself).

**Reply by rureverek** [reply]
Well, yes it's related to loading parameters of `ros2control`, not MoveIt. When placing `ros2control` in namespace `moto_gp4`, normally I would expect it to look for `controller_manager` in that namespace (`moto_gp4/controller_manager`). But it seems like I need to explicitely tell `ros2control` that my `controller_manager` is in `moto_gp4` namespace in ros2controllers.yaml.

I think I meant, that controller_manager is not visible for ros2control, not MoveIt in the post before... It's pretty easy to get lost here :)

**Reply by gavanderhoorn** [reply]
@rureverek: I've merged your comment with the launch file with your answer posted below.

> I think I meant, that controller_manager is not visible for ros2control, not MoveIt in the post before... It's pretty easy to get lost here :)

true.

Good to hear though my assumption about 'just placing MoveIt in a namespace' worked out, and did in fact reduce the nr of remaps to a minimum.

**Comment by rureverek** [comment, marked as answer]
**Update. @TE-2 handed over task to me. Finally, I managed to place two robots in two different moveIt nodes (namespaces). Here are my findings:**

I put all nodes in namespace e.g: robot_state_publisher,  run_move_group_node, ros2_control_node, joint_state_publisher.
Only one remapping was needed for robot_state_publisher (possibly could be fixed by using different source_list at joint_state_publisher, but I didn't verify that).

After that, two main problems came up when trying to run it with previous moveIt configuration:

**Problem: When placing controller_manager in namespace, controller_manager wasn't visible for ros2control.**

Solution:

Changed configuration: 
Add namespace prefix before controller_manager and moto_gp4 e.g moto_gp4/controller_manager, and moto_gp4/moto_gp4 in ros2_controllers_namespace.yaml:

```yaml
moto_gp4/controller_manager:
  ros__parameters:
    update_rate: 100  # Hz

    moto_gp4:
      type: joint_trajectory_controller/JointTrajectoryController


    joint_state_broadcaster:
      type: joint_state_broadcaster/JointStateBroadcaster


moto_gp4/moto_gp4:
	...
	...
```

I don't like that I had to put namespace prefix before the name of controller_manager in ros2_controllers_namespace.yaml, but apparrently it was the only way that worked for me. (I tried many simpler things, which didn't work).

**Error: "action client is not connected to action server".**

The key was to understand how moveit_controller.yaml works.
By default following configuration was used:
moveit_controllers.yaml:
```yaml
moveit_simple_controller_manager:
  controller_names:
    - moto_gp4

  moto_gp4:
    type: FollowJointTrajectory
    action_ns: follow_joint_trajectory
```

This is telling moveIt to look for action server in this path: "/moto_gp4/follow_joint_trajectory". However, its more problematic when we want to place MoveIt into new namespace... Then this "path" looks like this: "<moveit_ns>/<controller_name>/<action_ns>" (/moto_gp4/moto_gp4/follow_joint_trajectory). To solve this problem I changed the configuration of the controller manager running in namespace to following:

```yaml
moveit_simple_controller_manager:
  controller_names:
    - follow_joint_trajectory

  follow_joint_trajectory:
    type: FollowJointTrajectory
    action_ns: ""
```

In that case if we place the controller in namespace called "moto_gp4", we effectively end up with path: "/moto_gp4/follow_joint_trajectory", as action_ns is not defined here.

---

Edit: launch file for reference:

<details><summary>Click to expand</summary>

```py
import os
from launch import LaunchDescription
from launch.actions import DeclareLaunchArgument
from launch.substitutions import LaunchConfiguration
from launch.conditions import IfCondition, UnlessCondition
from launch_ros.actions import Node
from launch.actions import ExecuteProcess, Shutdown
from ament_index_python.packages import get_package_share_directory
from moveit_configs_utils import MoveItConfigsBuilder

import xacro


def generate_launch_description():

    moveit_namespace = "moto_gp4"
    moveit_namespace2 = "moto_mini"

    motoman_directory = "moveit_resources_moto_gp4_moveit_config"
    motoman_directory2 = "moveit_resources_moto_motomini_moveit_config"

    # URDF
    motoman_robot_description_config = xacro.process_file(
        os.path.join(
            get_package_share_directory(motoman_directory),
            "config",
            "motoman_gp4.urdf.xacro",
        ),
        # mappings={'prefix': '/yaskawa_a1/'},
    )
    motoman_robot_description = {
        "robot_description": motoman_robot_description_config.toxml()
    }

    # URDF
    motoman_robot_description_config2 = xacro.process_file(
        os.path.join(
            get_package_share_directory(motoman_directory2),
            "config",
            "motoman_motomini.urdf.xacro",
        ),
        # mappings={'prefix': '/yaskawa_a1/'},
    )
    motoman_robot_description2 = {
        "robot_description": motoman_robot_description_config2.toxml()
    }

    moveit_config = (
        MoveItConfigsBuilder("moveit_resources_moto_gp4")
        .robot_description(file_path="config/motoman_gp4.urdf.xacro")
        .robot_description_semantic(file_path="config/motoman_gp4.srdf")
        .trajectory_execution(file_path="config/moveit_controllers_shared.yaml")
        .to_moveit_configs()
    )

    moveit_config2 = (
        MoveItConfigsBuilder("moveit_resources_moto_motomini")
        .robot_description(file_path="config/motoman_motomini.urdf.xacro")
        .robot_description_semantic(file_path="config/motoman_motomini.srdf")
        .trajectory_execution(file_path="config/moveit_controllers_shared.yaml")
        .to_moveit_configs()
    )


    planning_scene_monitor_parameters = {
        "publish_planning_scene": True,
        "publish_geometry_updates": True,
        "publish_state_updates": True,
        "publish_transforms_updates": True,
        "publish_robot_description": True,
        "publish_robot_description_semantic": True,
    }
    # Start the actual move_group node/action server
    run_move_group_node = Node(
        package="moveit_ros_move_group",
        executable="move_group",
        output="screen",
        namespace=moveit_namespace,
        parameters=[
            moveit_config.to_dict(),
            planning_scene_monitor_parameters,
            # {"use_sim_time": True},
        ],
        # remappings=[
        #     ('joint_states', 'joint_state_broadcaster/joint_states')]
    )

    run_move_group_node2 = Node(
        package="moveit_ros_move_group",
        executable="move_group",
        output="screen",
        namespace=moveit_namespace2,
        parameters=[
            moveit_config2.to_dict(),
            planning_scene_monitor_parameters,
            # {"use_sim_time": True},
        ],
        # remappings=[
        #     ('joint_states', 'joint_state_broadcaster/joint_states')]
    )

    # RViz
    rviz_base = os.path.join(
        get_package_share_directory("moveit_resources_moto_gp4_moveit_config"), "launch"
    )
    rviz_full_config = os.path.join(rviz_base, "moveit_namespace.rviz")

    rviz_node = Node(
        package="rviz2",
        executable="rviz2",
        name="rviz2",
        output="log",
        arguments=["-d", rviz_full_config],
        parameters=[
            moveit_config.robot_description,
            moveit_config.robot_description_semantic,
            moveit_config.planning_pipelines,
            moveit_config.robot_description_kinematics,
            # moveit_config2.robot_description2,
            # {"use_sim_time": True},
        ],
    )

    # Publish TF
    robot_state_publisher = Node(
        package="robot_state_publisher",
        executable="robot_state_publisher",
        name="robot_state_publisher",
        output="both",
        namespace=moveit_namespace,
        parameters=[moveit_config.robot_description],
        remappings=[
        ('joint_states', 'joint_state_broadcaster/joint_states'),
            # ('/robot_description', f'/{moveit_namespace}/robot_description')
        ]
    )

    robot_state_publisher2 = Node(
        package="robot_state_publisher",
        executable="robot_state_publisher",
        name="robot_state_publisher",
        output="both",
        namespace=moveit_namespace2,
        parameters=[moveit_config2.robot_description],
        remappings=[
            ('joint_states', 'joint_state_broadcaster/joint_states')
        ]
    )

    # ros2_control using FakeSystem as hardware
    ros2_controllers_path = os.path.join(
        get_package_share_directory("moveit_resources_moto_gp4_moveit_config"),
        "config",
        "ros2_controllers_shared.yaml",
    )

    ros2_controllers_path2 = os.path.join(
        get_package_share_directory("moveit_resources_moto_motomini_moveit_config"),
        "config",
        "ros2_controllers_shared.yaml",
    )

    ros2_control_node = Node(
        namespace=moveit_namespace,
        package="controller_manager",
        executable="ros2_control_node",
        parameters=[moveit_config.robot_description, ros2_controllers_path], #Remember to add ros_control tag in xacro file
        output="both",
    )

    ros2_control_node2 = Node(
        namespace=moveit_namespace2,
        package="controller_manager",
        executable="ros2_control_node",
        parameters=[moveit_config2.robot_description, ros2_controllers_path2], #Remember to add ros_control tag in xacro file
        output="both",
    )


    joint_state_publisher = Node(
        package='joint_state_publisher',
        executable='joint_state_publisher',
        name='joint_state_publisher',
        namespace=moveit_namespace,
        parameters=[
            {
                "robot_description": motoman_robot_description,
                "rate": 43,
                "source_list": [f'/{moveit_namespace}/joint_states'],       # Könnte auch Moveit Problem sein
            }
        ],
    )

    joint_state_publisher2 = Node(
        package='joint_state_publisher',
        executable='joint_state_publisher',
        name='joint_state_publisher',
        namespace=moveit_namespace2,
        parameters=[
            {
                "robot_description": motoman_robot_description2,
                "rate": 43,
                "source_list": [f'/{moveit_namespace2}/joint_states'],       # Könnte auch Moveit Problem sein
            }
        ],
    )

    return LaunchDescription(
        [
            rviz_node,
            # static_tf,
            robot_state_publisher,
            robot_state_publisher2,
            run_move_group_node,
            run_move_group_node2,
            ros2_control_node,
            ros2_control_node2,
            joint_state_publisher,
            joint_state_publisher2
        ]
    )
```

</details>

**Reply by augustinoh** [reply]
Hi ! Is it possible to put a link to the whole folder with the code, because I am struggling to understand the whole configuration.

This would help me a lot understanding the whole picture.

Thank you in advance.
